qldpc/codes.py

```diff
@@ -675,6 +761,382 @@ def _minimize_weight_of_logical_op(
+            elif isinstance(code_z, ClassicalCode):
+                self.field_order = self.code_z._field_order
+            else:
+                self.field_order = DEFAULT_FIELD_ORDER
+
+        self.code_x = ClassicalCode(code_x, self.field_order)
+        self.code_z = ClassicalCode(code_z, self.field_order)
+        # if not self.code_x._field_order == self.code_z._field_order :
+        #    raise ValueError("The sub-codes provided for this CSSCode are over different fields")
+
+        if not self.code_x.num_bits == self.code_z.num_bits or np.any(
+            self.code_x.matrix @ self.code_z.matrix.T
+        ):
+            raise ValueError("The sub-codes provided for this CSSCode are incompatible")
+
+    @functools.cached_property
+    def matrix(self) -> npt.NDArray[np.int_]:
+        """Overall parity check matrix."""
+        matrix = np.block(
+            [
+                [np.zeros_like(self.code_z.matrix), self.code_z.matrix],
+                [self.code_x.matrix, np.zeros_like(self.code_x.matrix)],
+            ]
+        ).reshape((self.num_checks, 2, -1))
+        return matrix
+
+    @property
+    def num_checks(self) -> int:
+        """Number of parity checks."""
+        return self.code_x.matrix.shape[0] + self.code_z.matrix.shape[0]
+
+    @property
+    def num_qubits(self) -> int:
+        """Number of data qubits in this code."""
+        return self.code_x.matrix.shape[1]
+
+    @property
+    def num_logical_qubits(self) -> int:
+        """Number of logical qubits encoded by this code."""
+        return self.code_x.dimension + self.code_z.dimension - self.num_qubits
+
+    """
+    Code tranformation methods — Conjugation and Shifting.
+
+    (i) Conjugation — A CSSCode can specify which data qubits should be hadamard-transformed
+    before/after syndrome extraction, thereby transforming the operators
+    that address a specified data qubit as (X,Y,Z) <--> (Z,Y,X).
+
+    (ii) Shifting — A CSSCode can also "shift" the Pauli operators on a qubit, moving vertically
+    along the following table:
+
+    ―――――――――――
+    | XY | YX |
+    ―――――――――――
+    | XZ | ZX |
+    ―――――――――――
+    | YZ | ZY |
+    ―――――――――――
+
+    Qubit shifts are specified by a dictionary mapping qubit_index --> shift_index, where
+    shift_index = +1, 0, and -1 mod 3 respectively refer to the top, middle, and bottom rows of the
+    table.
+
+    Physically, a shift of +1 and -1 respectively correspond to conjugating the Pauli operators
+    addressing a qubit by sqrt(X) and sqrt(Z) rotations.
+    """
+
+    # Given a parity check matrix, swap X and Z operators on the qubits to conjugate.
+    # TODO: Does this work for general stabilizer code (over F_2)? If so, move to QuditCode class?
+    def conjugate(cls, matrix: IntegerMatrix | None, conj: slice | Sequence[int]) -> IntegerMatrix:
+        matrix[:, :, conj] = np.roll(matrix[:, :, conj], 1, axis=1)
+        return matrix
```

Comment:
I think this one is general.  The shift below is not.

Reply:
Good idea!